docs/coding.md

```diff
@@ -35,3 +35,81 @@ new Rust file types.
 [file]: https://shadow.github.io/docs/rust/shadow_rs/host/descriptor/enum.File.html
 [descriptor]: https://shadow.github.io/docs/rust/shadow_rs/host/descriptor/struct.Descriptor.html
 [dup]: https://shadow.github.io/docs/rust/shadow_rs/host/descriptor/struct.Descriptor.html#method.dup
+
+## Platform (libc and Linux) crates
+
+We use several Rust crates for accessing platform functionality and definitions.
+Roughly from lowest-level to highest-level:
+
+* Our [`linux-api`](https://github.com/shadow/shadow/tree/main/src/lib/linux-api)
+crate provides fairly low-level bindings over the Linux kernel headers, and a
+few `nix`-style higher-level wrappers. It does not depend on `std` or `libc`.
+It also re-exports these definitions as a C library that can be used without
+creating conflicts with libc headers or linux system headers.
+Use this when working with the syscall ABI (such as when implementing syscall
+handlers), and for making syscalls when none of the higher-level crates are
+suitable (see below).
+
+*   [`libc`](https://docs.rs/libc/latest/libc/) provides fairly low-level bindings
+of system libc standard headers. If you need syscall-level ABI-compatibility,
+use `linux-api` instead. If you don't, prefer one of the higher-level crates.
+It's sometimes useful when delegating the implementation of a syscall to the
+host kernel, though sometimes the `libc` interfaces don't quite match the
+syscall functionality we're actually trying to implement.
```

Comment:
This sentence isn't clear to me.

Reply:
expanded:

e.g. suppose we want to emulate
a `fcntl` syscall by executing that syscall ourselves from shadow. Calling `libc`'s
`fcntl` function with the syscall parameters is *probably* equivalent to calling the `fcntl`
syscall directly, but it's difficult to be certain that the libc wrapper doesn't
use a separate set of of "command" constants and remap internally to the syscall constants.

Reply:
Okay I see. I don't know if you have any options in mind, but one example where this does happen is `SO_TIMESTAMP`, which libc maps to either `SO_TIMESTAMP_OLD` or `SO_TIMESTAMP_NEW` when making the syscall.

Reply:
Hmm, it's a neat example, but a little weird in that it wouldn't make sense for shadow to make native `setsockopt` calls.

I tweaked this a bit more and moved it to the section below about preferring `linux-api` for syscall handlers.